src/impl_methods.rs

```diff
@@ -2231,4 +2231,60 @@ where
             })
         }
     }
+
+    /// Iterates over pairs of consecutive elements along the axis.
+    ///
+    /// The first argument to the closure is an element, and the second
+    /// argument is the next element along the axis. Iteration is guaranteed to
+    /// proceed in order along the specified axis, but in all other respects
+    /// the iteration order is unspecified.
+    ///
+    /// # Example
+    ///
+    /// For example, this can be used to compute the cumulative sum along an
+    /// axis:
+    ///
+    /// ```
+    /// use ndarray::{array, Axis};
+    ///
+    /// let mut arr = array![
+    ///     [[1, 2], [3, 4], [5, 6]],
+    ///     [[7, 8], [9, 10], [11, 12]],
+    /// ];
+    /// arr.accumulate_axis_inplace(Axis(1), |&prev, curr| *curr += prev);
+    /// assert_eq!(
+    ///     arr,
+    ///     array![
+    ///         [[1, 2], [4, 6], [9, 12]],
+    ///         [[7, 8], [16, 18], [27, 30]],
+    ///     ],
+    /// );
+    /// ```
+    pub fn accumulate_axis_inplace<F>(&mut self, axis: Axis, mut f: F)
```

Comment:
This method is a neat demonstration of an array Zip-ing with itself. It seems a bit niche but we can of course include it. But the question of course goes out, how do we generalize this. Apart from the raw pointer Zip, of course?

Reply:
I think we really need to wait until Rust has generic associated types before we can write good self-Zip abstractions. We can do things like this today:

```rust
    /// Folds over the mutable windows of the specified size.
    pub fn windows_mut_fold<E, B, F>(&mut self, window_size: E, init: B, f: F)
    where
        E: IntoDimension<Dim = D>,
        F: for<'a> FnMut(B, ArrayViewMut<'a, A, D>) -> B,
        S: DataMut,
    {
        // ...
    }
```

but without generic associated types, we can't abstract this into a `WindowsMut` producer that implements a trait, because the item type (which is an associated type of the trait) has a temporary lifetime.